### PR TITLE
Allow to override the CustomerFormatter class

### PR DESCRIPTION
**Pull request by rGaillard**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop 
| Description?  | This is currently not possible to override the CustomerFormatter class because of the private properties
 Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11910)

### CODE REVIEW

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Look legit, thanks @rGaillard

**Comment by mickaelandrieu** [comment]
We don't want to allow the override anymore, this is why you are free to implement the interface and adapt your code to fits your specific needs.

Can you tell us more about it?

File: classes/form/CustomerFormatter.php [hunk 1]
```diff
@@ -27,15 +27,15 @@
 
 class CustomerFormatterCore implements FormFormatterInterface
 {
-    private $translator;
-    private $language;
+    protected $translator;
```
Comment (mickaelandrieu) [inline classes/form/CustomerFormatter.php:30]: this is not something we want to open, you can alter it by injecting your own in the constructor

File: classes/form/CustomerFormatter.php [hunk 2]
```diff
@@ -247,7 +247,7 @@ public function getFormat()
         return $this->addConstraints($format);
     }
 
-    private function addConstraints(array $format)
+    protected function addConstraints(array $format)
```
Comment (mickaelandrieu) [inline classes/form/CustomerFormatter.php:250]: same here, this function is not supposed to be called by your user, but in your own implementation you can do everything you want + and have to maintain it +. The main key is here, and a very important shift compared to 1.6- history

**Review by mickaelandrieu — CHANGES_REQUESTED** [review]

File: classes/form/CustomerFormatter.php (hunk 2 above)
Comment (kpodemski) [reply classes/form/CustomerFormatter.php:250]: what you mean "your own implementation"? if you need to do custom registration form with altered fields, username as a login and custom fields with custom validation what you would do?

**Comment by rGaillard** [comment]
> We don't want to allow the override anymore, this is why you are free to implement the interface and adapt your code to fits your specific needs.
> 
> Can you tell us more about it?

I just want to change the behavior and the order of some fields in my custom form and i don't want to copy/paste everything in a new formatter for that... this is a no sense! Also, if tomorrow a bug is fixed, it will not be fixed for me...
Currently, i have 2 choice:
- implement my own Formatter with a copy/past from the original...
- re-implement the constructor (for the translator and the language attributes) and the addConstraint method in my override...

Removing overrides increase a lot the development and the maintenance time for any project with specific needs....

File: classes/form/CustomerFormatter.php (hunk 2 above)
Comment (rGaillard) [reply classes/form/CustomerFormatter.php:250]: @mickaelandrieu why should i do a copy/paste? I just want to use this identical method in my override! Why should i maintaining myself something that is already maintained? This is just a waste of time...

File: classes/form/CustomerFormatter.php (hunk 1 above)
Comment (rGaillard) [reply classes/form/CustomerFormatter.php:30]: @mickaelandrieu I don't want to alter it, i just want to use it... And if i wanted to alter it, what is the problem?

**Comment by mickaelandrieu** [comment]
> I just want to change the behavior

This is why you shouldn't rely on the current implementation and why private functions (ie outside of public api) shouldn't be put in protected visibility.

I'm totally fine with you guys copy pasting code and maintain your own code, I'm not fine with you guys complaining that core classes have changed and break "your code" during an auto upgrade.

Projects with specific needs should have specific implementations: as developers, it's up to you to decide if the ones provided by the core fits your needs or not. As core team member, it's mine to decide if we maintain more code on our own for your specific needs.

This is not something we want to do right now, it's far better to copy paste and adapt than extends a class you're supposed to override.

>  if you need to do custom registration form with altered fields, username as a login and custom fields with custom validation what you would do?

Except if I'm wrong, the best and recommended way to alter forms is to use hooks: it's even easier and more powerful on migrated pages.

In fact, this is why I asked more information : maybe a hook here could be useful, in addition of already provided hooks to manipulate forms rendering, validation and submission

File: classes/form/CustomerFormatter.php (hunk 1 above)
Comment (mickaelandrieu) [reply classes/form/CustomerFormatter.php:30]: I don't have any problem with it, in your own implementation. This class is a CustomerFormatter and not supposed to open a translator instance to users.

**Comment by eternoendless** [comment]
Overrides should be avoided as much as possible. PrestaShop is moving away from an inheritance-based extension system to a composition-based one.

The O in SOLID stands for "Open/Closed Principle" which states that "software entities should be open for extension, but closed for modification". 

In that scenario, behavior of an entity can be changed by either:
- Constructing the element using dependencies that behave differently than the ones before
- Replacing the object with another implementation that follows the same interface

Yes, this is tricky and maybe a little more work, especially since not all of PrestaShop's code is currently SOLID. But it will give your code better chances at surviving an upgrade later.

PS: Some "copy/paste" cases can be solved using a Decorator that implements the same interface as the class it's replacing.

**Comment by Quetzacoalt91** [comment]
Thank for contributing to this discussion. Regarding the last messages, I'm closing this PR as it seems you can easily find a workaround for overriding the class, even if it requires some/copy paste for the properties.

**Pull request closed without merging**